Adding remove-feature to Cron-Class

HpLightcorner opened the pull request:
Adding functions to remove a specific schedule (by the given name) or all scheduled tasks from the Cron class.

I was missing a feature to remove a task once it was scheduled. So I added some functions to handle my request. Editing the tasks is thread-safe: The queue can only be manipulatet when `tick` is not called.

PerMalmberg reviewed (changes requested):
Hello and thank you for the contribution :)

I've left some comments throughout the PR, but please read the following before running off addressing them.

I'm not against the idea of removing/clearing the scheduler, it's a nice addition. I do however object against making the scheduler thread-aware for the following reasons:

1. Any user of this library would, with this change, pay the cost of locking/unlocking the mutex _each tick_, even if the application itself is not multi threaded.
1. The scheduler itself does not require thread-awareness and as such it should not be. Such logic can easily be handled in the application itself; either use a `std::unique_lock` whenever accessing the scheduler, or put the scheduler in a thread-aware wrapper to abstract the thread-handling away from the application.

Side note: if removing/clearing schedules requires locking, then so should `add_schedule`, that's currently missing.

So, I will not accept the PR with the thread-awareness; the clearing/removing functionality is however welcome.

HpLightcorner commented:
Hey,
I see your points and completely agree with you. What do you think about the following:
1. We could exten the Cron class by an additiona template paramter, the 'TaskLockType'
2. We then define two alias for Cron, 'Cron' and 'CronMt'
3. We add the Thread-Safety to 'add_schedule'

Let's have a look at this pseudo-code to make my idea more clear:
```
    class TaskLockerNone 
    {
        public:
            void lock() {}
            void unlock() {}
    };

    class TaskLocker
    {
        public:
            TaskLocker() : lock_(m, std::defer_lock) {}
            void lock() { lock_.lock(); }
            void unlock() { lock_.unlock(); }
        private:
            std::mutex m{};
            std::unique_lock<std::mutex> lock_;
    };

    template<typename ClockType, typename TaskLockType>
    class CronBase
   {

   private:
            class Queue
                    // Priority queue placing smallest (i.e. nearest in time) items on top.
                    : public std::priority_queue<Task, std::vector<Task>, std::greater<>>
            {
                public:
                    // Inherit to allow access to the container.
                    const std::vector<Task>& get_tasks() const
                    {
                        return c;
                    }

                    std::vector<Task>& get_tasks()
                    {
                        return c;
                    }

                    void clear()
                    {
                        lock.lock();

                        while (!empty())
                            pop();

                        lock.unlock();
                    }

                    template<typename T>
                    void remove(T to_remove)
                    {
                        lock.lock();

                        /* Copy current elements */
                        std::vector<Task> temp{};
                        std::swap(temp, c);

                        /* Refill with elements ensuring correct order by calling push */
                        for (const auto& task : temp)
                        {
                            if (task != to_remove)
                                push(task);
                        }

                        lock.unlock();
                    }

                    void lock_queue()
                    {
                        /* Do not allow to manipulate the Queue */
                        lock.lock();
                    }

                    void release_queue()
                    {
                        /* Allow Access to the Queue Manipulating-Functions */
                        lock.unlock();
                    }
                    
                private:
                    TaskLockType lock;
            };
};

    template<typename ClockType = libcron::LocalClock>
    using Cron = CronBase<ClockType, libcron::TaskLockerNone>;

    template<typename ClockType = libcron::LocalClock>
    using CronMt = CronBase<libcron::LocalClock, libcron::TaskLocker>;

```

Hence the User can decide for Multithread-Support. I think the compiler will take care to optimize the empty function calls?

PerMalmberg commented:
That's a solution I can live with, but I've got to ask; is there a reason you need it in the library and can't just extend it in your own project?

Make the new template argument be the first one and non-locking by default such that the current API is unchanged, i.e. no code changes need when updating. And no warts in variable names please (`lock_`).

HpLightcorner commented:
For example, in my application I have multiple instances of Cron used in multiple threads where each thread shall be able to add or remove a schedule - however, only a single thread is handling timing and calling the tick function (a high precision timer implementation). So, I think it would somehow inconvenient to take care of Multithreading on my own as the Cron class exactly knows when to make sure that the Cron-Tasks are protected - and as a user of the library I would at least prefer if the library offers me a convenient way to handle this crucial part.

That said, I think it should be the libraries responsibility to inform users at least about possible pitfalls or come up with a solution. So, when not providing a thread safe interface (like CronMt), it would at least suggest adding a corresponding warning/documentation.

As CronMt is a solution you can live with, I will adjust my code so that you can finish the pull request!

PerMalmberg commented:
Leave the aliases outside of the library - they are for the user to decide the naming of. It's enough that the library provides the means via the template parameters. I'd rather just give an example in the docs.

HpLightcorner commented:
Hey, 
I added the changes as discussed, however, to not break the API, I think the `TaskLocker` has to be the last/second template parameter. I also added some lines of documentation and tests.

PerMalmberg commented on libcron/include/libcron/Cron.h:
```diff
@@ -4,23 +4,44 @@
 namespace libcron
 {
-    template<typename ClockType>
+    class TaskLockerNone 
```
```suggestion
    class NullLock 
```

PerMalmberg commented on libcron/include/libcron/Cron.h:
```diff
@@ -4,23 +4,44 @@
+            void unlock() {}
+    };
+
+    class TaskLocker
```
```suggestion
    class Locker
```

PerMalmberg commented on libcron/include/libcron/Cron.h:
```diff
@@ -4,23 +4,44 @@
 
-    template<typename ClockType>
-    std::ostream& operator<<(std::ostream& stream, const Cron<ClockType>& c);
+    template<typename ClockType, typename TaskLockType>
```
```suggestion
    template<typename ClockType, typename LockType>
```
LockType says what it does, no need for the extra 'Task'. Need to change this throughout the file too.

PerMalmberg commented on README.md:
````diff
@@ -1,7 +1,67 @@
+When Calling `libcron::Cron::tick` from another thread than `add_schedule`, `clear_schedule` and `remove_schedule`, one must take care to protect the internal resources of `libcron::Cron` so that tasks are not removed or added while `libcron::Cron` is iterating over the schedules. `libcron::Cron` can take care of that, you simply have to define your own aliases:
+
+```
+/* The defaut class uses TaskLockerNone, which does no lock the ressources at runtime */
````
```suggestion
/* The default class uses NullLock, which does not lock the resources at runtime */
```

PerMalmberg reviewed (changes requested):
Except a few naming items it looks good.

HpLightcorner commented:
Done!

PerMalmberg commented:
Merged!
